bin/propolis-server/src/lib/migrate/destination.rs

```diff
@@ -24,105 +27,321 @@ use crate::migrate::probes;
 use crate::migrate::{
     Device, MigrateError, MigratePhase, MigrateRole, MigrationState, PageIter,
 };
-use crate::vm::migrate_commands::MigrateTargetCommand;
-use crate::vm::migrate_commands::MigrateTargetResponse;
+use crate::vm::ensure::{VmEnsureActive, VmEnsureNotStarted};
+use crate::vm::objects::VmObjectsShared;
+use crate::vm::state_publisher::{ExternalStateUpdate, MigrationStateUpdate};
 
 use super::protocol::Protocol;
+use super::MigrateConn;
+
+/// The interface to an arbitrary version of the target half of the live
+/// migration protocol.
+//
+// Use `async_trait` here to help generate a `Send` bound on the futures
+// returned by the functions in this trait.
+#[async_trait::async_trait]
+pub(crate) trait DestinationProtocol<'e> {
+    /// Runs live migration as a target, attempting to create a set of VM
+    /// objects in the process. On success, returns an "active VM" placeholder
+    /// that the caller can use to set up and start a state driver loop.
+    async fn run<'ensure>(
+        mut self,
+        ensure: VmEnsureNotStarted<'ensure>,
+    ) -> Result<VmEnsureActive<'ensure>, MigrateError>
+    where
+        'e: 'ensure;
+}
 
-/// Launches an attempt to migrate into a supplied instance using the supplied
-/// source connection.
-pub async fn migrate<T: AsyncRead + AsyncWrite + Unpin + Send>(
+/// Connects to a live migration source using the migration request information
+/// in `migrate_info`, then negotiates a protocol version with that source.
+/// Returns a [`DestinationProtocol`] implementation for the negotiated version
+/// that the caller can use to run the migration.
+pub(crate) async fn initiate<'e>(
```

Comment:
This function is more or less identical to the old `dest_initiate` function in the root migration module.